xdsl/irdl/constraints.py

```diff
@@ -465,12 +466,18 @@ def __init__(
 
         bases = set[Attribute]()
         eq_bases = set[Attribute]()
+        abstr_constr: AttrConstraint[AttributeCovT] | None = None
         for i, c in enumerate(constrs):
             b = c.get_bases()
             if b is None:
-                raise PyRDLError(
-                    f"Constraint {c} cannot appear in an `AnyOf` constraint as its bases aren't known."
-                )
+                if abstr_constr is None:
+                    abstr_constr = c
+                    continue
+                else:
+                    raise PyRDLError(
+                        "Only one abstract constraint is allowed in `AnyOf` constraint,"
+                        f" found {c} when {abstr_constr} was already present."
+                    )
```

Comment:
we have to check here that none of the bases or eq bases satisfy the abstract constraint